go/stats/counters.go

```diff
@@ -19,36 +19,99 @@ package stats
 import (
 	"bytes"
 	"fmt"
+	"strconv"
 	"strings"
 	"sync"
 	"sync/atomic"
+	"time"
+
+	"vitess.io/vitess/go/vt/logutil"
+
+	"vitess.io/vitess/go/sync2"
 )
 
+// logCounterNegative is for throttling adding a negative value to a counter messages in logs
+var logCounterNegative = logutil.NewThrottledLogger("StatsCounterNegative", 1*time.Minute)
+
+// Counter is expvar.Int+Get+hook
+type Counter struct {
+	i    sync2.AtomicInt64
+	help string
+}
+
+// NewCounter returns a new Counter
+func NewCounter(name string, help string) *Counter {
+	v := &Counter{help: help}
+	if name != "" {
+		publish(name, v)
+	}
+	return v
+}
+
+// Add adds the provided value to the Counter
+func (v *Counter) Add(delta int64) {
+	if delta < 0 {
+		logCounterNegative.Warningf("Adding a negative value to a counter, %v should be a gauge instead", v)
+	}
+	v.i.Add(delta)
+}
+
+// Reset resets the counter value to 0
+func (v *Counter) Reset() {
+	v.i.Set(int64(0))
+}
+
+// Get returns the value
+func (v *Counter) Get() int64 {
+	return v.i.Get()
+}
+
+// String is the implementation of expvar.var
+func (v *Counter) String() string {
+	return strconv.FormatInt(v.i.Get(), 10)
+}
+
+// Help returns the help string
+func (v *Counter) Help() string {
+	return v.help
+}
+
+// Gauge is an unlabeled metric whose values can go up/down.
+type Gauge struct {
+	Counter
+}
+
+// NewGauge creates a new Gauge and publishes it if name is set
+func NewGauge(name string, help string) *Gauge {
+	v := &Gauge{Counter: Counter{help: help}}
+
+	if name != "" {
+		publish(name, v)
+	}
+	return v
+}
+
+// Set sets the value
+func (v *Gauge) Set(value int64) {
+	v.Counter.i.Set(value)
+}
+
+// Add adds the provided value to the Gauge
+func (v *Gauge) Add(delta int64) {
+	v.Counter.i.Add(delta)
+}
+
 // Counters is similar to expvar.Map, except that
-// it doesn't allow floats. In addition, it provides
-// a Counts method which can be used for tracking rates.
+// it doesn't allow floats. It is used to build CountersWithLabels and GaugesWithLabels.
 type Counters struct {
```

Comment:
Is it actually necessary to export this?

I saw that you're using it in the promstatsbackend but the usage was not immediately clear to me.

I'm asking because we have several type switches in the code base and I made the mistake of using "stats.Counter" instead of "stats.CountersWithLabels" in these type switches.

Reply:
Yeah I agree, it doesn't make sense. Let me subsume what it's doing into CountersWithLabels & GaugesWithLabels.